### PR TITLE
Adding support to parser for a few more constructs

### PR DESCRIPTION
**Pull request by bbeaudreault**
I'm going to add a few more tests specifically for the empty `values()` list, and the change to IsValue and AsInterface. But I think the changes should be good to review whenever you have time @sougou 

- DATE Function
- LIKE pattern ESCAPE char syntax
- && as AND
- Charset Introducers on integral values
- pound-sign comments
- INSERT with no values

For insert with no values, this is to support cases where there is a table where all columns have defaults and auto_inc.  Here's the docs on charset introducers and LIKE ESCAPE:

https://dev.mysql.com/doc/refman/5.7/en/charset-introducer.html
https://dev.mysql.com/doc/refman/5.7/en/string-comparison-functions.html#operator_like

### CODE REVIEW

**Comment by bbeaudreault** [comment]
Added some tests. Ready for review

**Comment by bbeaudreault** [comment]
Test failures are legit, and due to my special treatment of sql id's that start with an underscore. Looking into it.

**Comment by bbeaudreault** [comment]
@sougou This PR is now a bit more trivial, and is ready for review.

I removed the Charset Introducer support because of shift/reduce conflicts. MySQL gets around this by checking for the existence of the charset loaded into the server when deciding if a token starting with _ is a charset introducer or a normal identifier. Since we don't have that knowledge in VTGate, we can't differentiate between a (reduce) of as_ci_opt and a (shift) into the charset introducer.

We're going to see if we can just stop using charset introducers instead of support it. I kept the commit and revert around for now in case we need a history of the attempt, but if you'd like I can squash and keep that history in HubSpot's fork instead. We may need to come down this road again.

**Comment by bbeaudreault** [comment]
This became mildly un-trivial again.  But I think it does a lot to move us closer to parity with mysql. What I added:

- Refactored expression and the tree of rules under it -- it now much more closely mirrors [mysql's yacc grammar](https://github.com/mysql/mysql-server/blob/5.7/sql/sql_yacc.yy) at least in structure, for this part (search `expr:` or `bool_pri:` and step through the 2 side-by-side). This makes it much easier to debug differences between the 2 parsers, because we're comparing common names and general structure.

- Add support for KEYWORD's as normal identifiers.  MySQL actually gets a little further on this than I did, and I'm not sure how yet. But this is a really good start and I think we can build upon it if we find more unparseable queries. For now we can use reserved keywords without backticks in qualified names, per the documentation:

> Exception: A word that follows a period in a qualified name must be an identifier, so it need not be quoted even if it is reserved

- As part of the above, I unified `TableIdent` and `ColIdent` onto just `Ident`. The main differences seemed to be that TableIdent is case-sensitive. For now I handled that by using a `CaseSensitiveEqual` function when comparing tables.

The resulting grammar has no conflicts, and I commented on areas where if we wanted to add support for constructs MySQL supports we can inject the rules in the right spot. It does *not* get us 100% coverage of keywords, but it gets us pretty close. For instance, MySQL can handle a reserved keyword in the first part of a qualified name. So where they support `reserved_sql_id '.' reserved_sql_id`, I was only able to support `sql_id '.' reserved_sql_id`.  I think we could work on pushing that further in a separate PR if necessary.

**Comment by sougou** [comment]
I eyeballed the sql.y changes. A few initial reactions:
* I want to us to avoid copying the MySQL grammar, mainly because we have the opportunity to do it right or better. If we find no other way, then we can do what they did :). In this light, which construct prompted you to change the expressions grammar?
* There is value in keeping ColIdent and TabIdent different. This allows us to rely on the compiler from someone forgetting about their case-sensitivity and performing the wrong kind of comparison. I want to see if we can still preserve that difference.
* I actually don't know why I never returned the bytes that represented keywords. I need to think if there was some performance related concern.

It's likely that I'll agree with these changes once I understand the root cause and think through our options. But for that, I need to dedicate more time.

**Comment by bbeaudreault** [comment]
Thanks for taking a look @sougou. After thinking about it more I actually left a note for myself last night as I went to bed to revert the ColIdent part of the change today. I agree with you and also I think there are cases where a table factor and select expression should use different rules for things like sql id. That may also bring us closer to supporting more keywords in select expressions. 

In terms of why the refactor: I was trying to figure out how some keywords that weren't explicitly listed were being allowed as identifiers for MySQL. I was looking at their yacc and trying to compare the expression grammars and having a hard time making heads or tails of the ways in which we differed. Good point on not assuming they did everything right. However I do like their hierarchy for expression. We were doing similar but it goes a little further in splitting rules based on their logical groupings. I also think it's a little more flexible based on what it allows, but I still need to write those tests. 

I'm totally on board for aiming for simplicity and performance over perfect semantic correctness, where we don't need it for the work vitess needs to do. Otherwise let them do the validation. 

But I also think over time I'll be needing to add more rules here because hubspot has a wide range of people doing a wide range of queries. I think it'll help to model the structure a bit where possible, if not the actual grammar rules. This will provide some point of reference when looking at how MySQL handles a particular grammar. 

In terms of performance of always returning bytes, I thought about that but wasn't sure. Do we have a benchmark we could run? The alternative would be to modify the keywords grammar to hard code returning the proper string for all the various keywords. I think the percentage of time keywords are used as identifiers will be much less than the percentage of time they are used as real keywords where the bytes are ignored. So that would probably be more performant.

**Comment by bbeaudreault** [comment]
Ok I reverted the ColIdent/TableIdent change. In order to avoid grammar conflicts, I kept `sql_id` in the qualifer part of column_name and the wildcard syntax of select_expression. But I created a `ColIdent.AsTableIdent()` to ensure that we pass the right type into the return values of those grammars.

**Comment by bbeaudreault** [comment]
Test failure seems to be due to mysql connection issues in travis. Will be pushing more tests today, which will cause a re-run.

**Comment by bbeaudreault** [comment]
Added support for parsing identifiers that are part of the UNUSED keyword category. I think that must have been lost somewhere along the way, as I see that was part of the point of including that map in the first place.

**Comment by bbeaudreault** [comment]
I turned sqlparser.TestValid into a BenchmarkValid and ran a 60s benchmark 3 times against each of master and this branch.  The average ns/op across all runs is about 70k ns/op slower in this PR than in master. It was 3843942 ns/op vs 3914577 ns/op.  That's about 1.8% difference.

This is not a very scientific test, as I'm running it on my local computer with other programs up (but while I'm away), and there are a bunch of different changes in this PR including the refactor and the new returning of bytes for everything.

I'm going to quickly see if it improves at all by not returning bytes for everything.  Not sure what your thoughts are on this level of impact.

**Comment by sougou** [comment]
Thanks for the benchmarks. 1.8% is negligible. We can go forward.

**Comment by bbeaudreault** [comment]
Great, thanks @sougou

**Comment by bbeaudreault** [comment]
Now that @acharis's PR was merged I'm going to rebase this. I'm also going to simplify my changes. To your previous point, I don't think the full refactor of the expr rules is necessary right now. So I'm going to back that part out, ensure the changes I want are still in there, and then rebase that cleaned up version.

**Comment by bbeaudreault** [comment]
@sougou ok I removed part of the refactor and rebased on master. This is ready for review.

File: go/vt/sqlparser/sql.y [hunk 1]
```diff
@@ -433,13 +439,13 @@ select_expression:
   {
     $$ = &NonStarExpr{Expr: $1, As: $2}
   }
-| table_id '.' '*'
+| sql_id '.' '*'
```
Comment (bbeaudreault) [inline go/vt/sqlparser/sql.y]: This and the other instance of AsTableIdent below (in column_name) are to avoid conflicts with things like function calls and other sql_ids

Comment (bbeaudreault) [inline go/vt/sqlparser/sql.y]: Actually I found a cleaner way around this, as you can now see in the latest push

File: go/vt/sqlparser/ast.go [hunk 2]
```diff
@@ -1032,6 +1033,9 @@ const (
 // Format formats the node.
 func (node *ComparisonExpr) Format(buf *TrackedBuffer) {
 	buf.Myprintf("%v %s %v", node.Left, node.Operator, node.Right)
+	if node.Escape != nil {
+		buf.Myprintf(" escape %v", node.Escape)
+	}
```
Comment (sougou) [inline go/vt/sqlparser/ast.go]: WalkSubtree needs to cover this new node. You may also have to verify that it doesn't crash if the expression is nil.

File: go/vt/sqlparser/ast.go [hunk 3]
```diff
@@ -1463,9 +1467,10 @@ func (node *CollateExpr) WalkSubtree(visit Visit) error {
 
 // FuncExpr represents a function call.
 type FuncExpr struct {
-	Name     ColIdent
-	Distinct bool
-	Exprs    SelectExprs
+	Qualifier TableIdent
```
Comment (sougou) [inline go/vt/sqlparser/ast.go]: Add to WalkSubtree.

File: go/vt/sqlparser/ast.go (hunk 2 above)
Comment (bbeaudreault) [inline go/vt/sqlparser/ast.go]: Added to WalkSubtree. Not sure what you mean by verifying that a nil expression doesn't crash it. I checked Walk and it does a nil-check. In terms of a nil expression in Format, it'd be a syntax error to have a ComparisonExpr without a left or right but with an escape.

File: go/vt/sqlparser/ast.go (hunk 3 above)
Comment (bbeaudreault) [inline go/vt/sqlparser/ast.go]: Added

**Comment by sougou** [comment]
LGTM
